### PR TITLE
gh-126925: Modify how iOS test results are gathered

### LINKED ISSUE #126821: Add option for redirecting stdout and stderr to the system log
# Feature or enhancement

### Proposal:

Python currently assumes that stdout and stderr are streams that are universally available. This is true for most systems. 
However, on iOS, stdout/stderr content is only really visible while in Xcode. When an app is running in the wild, the expectation is that the app will use the system log for all output needs. 

On iOS, stdout and stderr should be automatically directed to the system log (with appropriate log levels). This matches what Android does by default (see #118063). Existing iOS apps in the wild need to use a shim such as [std-nslog](https://pypi.org/project/std-nslog/) to ensure that stdout/err content can be loaded at runtime. This also complicates the iOS testing story, as stdout/stderr content is only visible at the end of a test run, rather than being streamable content as the test runs.

As iOS and macOS share an underlying logging framework, it should also be possible to have this functionality on macOS as well (on an opt-in basis when the Python interpreter is configured). This would be useful for Python code that has been embedded into a native macOS GUI app. At present, stdout and stderr are essentially inaccessible for apps of this type.

### Has this already been discussed elsewhere?

This is a minor feature, which does not need previous discussion elsewhere

### Links to previous discussion of this feature:

_No response_

### Linked PRs
* gh-127755
* gh-127806

### LINKED ISSUE #126925: "make testios" fails with Xcode 16+ due to changes in xcresulttool
As of Xcode 16, Apple has changed the command interface to `xcresultool` such that the way it is invoked when running the `testios` recipe in `Makefile.pre.in` is no longer valid, causing the iOS test step to fail while attempting to extract the results of the test run.

```pytb
[...]
# Regardless of success or failure, extract and print the test output
xcrun xcresulttool get --path iOSTestbed.arm64-iphonesimulator.1731824398/arm64-iphonesimulator.xcresult \
		--id $( xcrun xcresulttool get --path iOSTestbed.arm64-iphonesimulator.1731824398/arm64-iphonesimulator.xcresult --format json | _PYTHON_PROJECT_BASE=/Users/nad/Projects/PyDev/active/dev/3x/source/build-arm64-apple-ios-simulator _PYTHON_HOST_PLATFORM=ios-13.0-arm64-iphonesimulator PYTHONPATH=../Lib _PYTHON_SYSCONFIGDATA_NAME=_sysconfigdata__ios_arm64-iphonesimulator _PYTHON_SYSCONFIGDATA_PATH=/Users/nad/Projects/PyDev/active/dev/3x/source/build-arm64-apple-ios-simulator/build/lib.ios-13.0-arm64-iphonesimulator-3.14 /Users/nad/Projects/PyDev/active/dev/3x/source/build-arm64-apple-darwin/root-arm64-apple-darwin/bin/python3 -c "import sys, json; result = json.load(sys.stdin); print(result['actions']['_values'][0]['actionResult']['logRef']['id']['_value'])" ) \
		--format json | \
		_PYTHON_PROJECT_BASE=/Users/nad/Projects/PyDev/active/dev/3x/source/build-arm64-apple-ios-simulator _PYTHON_HOST_PLATFORM=ios-13.0-arm64-iphonesimulator PYTHONPATH=../Lib _PYTHON_SYSCONFIGDATA_NAME=_sysconfigdata__ios_arm64-iphonesimulator _PYTHON_SYSCONFIGDATA_PATH=/Users/nad/Projects/PyDev/active/dev/3x/source/build-arm64-apple-ios-simulator/build/lib.ios-13.0-arm64-iphonesimulator-3.14 /Users/nad/Projects/PyDev/active/dev/3x/source/build-arm64-apple-darwin/root-arm64-apple-darwin/bin/python3 -c "import sys, json; result = json.load(sys.stdin); print(result['subsections']['_values'][1]['subsections']['_values'][0]['emittedOutput']['_value'])"
Error: This command is deprecated and will be removed in a future release, --legacy flag is required to use it.
Usage: xcresulttool get object [--legacy] --path <path> [--id <id>] [--version <version>] [--format <format>]
  See 'xcresulttool get object --help' for more information.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
    import sys, json; result = json.load(sys.stdin); print(result['actions']['_values'][0]['actionResult']['logRef']['id']['_value'])
                               ~~~~~~~~~^^^^^^^^^^^
  File "/Users/nad/Projects/PyDev/active/dev/3x/source/Lib/json/__init__.py", line 293, in load
    return loads(fp.read(),
        cls=cls, object_hook=object_hook,
        parse_float=parse_float, parse_int=parse_int,
        parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "/Users/nad/Projects/PyDev/active/dev/3x/source/Lib/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
           ~~~~~~~~~~~~~~~~~~~~~~~^^^
  File "/Users/nad/Projects/PyDev/active/dev/3x/source/Lib/json/decoder.py", line 345, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
               ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/nad/Projects/PyDev/active/dev/3x/source/Lib/json/decoder.py", line 363, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
Error: Missing value for '--id <id>'
Help:  --id <id>  The ID of the object [optional, assumes rootID if not specified].
Usage: xcresulttool get object [--legacy] --path <path> [--id <id>] [--version <version>] [--format <format>]
  See 'xcresulttool get object --help' for more information.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
    import sys, json; result = json.load(sys.stdin); print(result['subsections']['_values'][1]['subsections']['_values'][0]['emittedOutput']['_value'])
                               ~~~~~~~~~^^^^^^^^^^^
  File "/Users/nad/Projects/PyDev/active/dev/3x/source/Lib/json/__init__.py", line 293, in load
    return loads(fp.read(),
        cls=cls, object_hook=object_hook,
        parse_float=parse_float, parse_int=parse_int,
        parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "/Users/nad/Projects/PyDev/active/dev/3x/source/Lib/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
           ~~~~~~~~~~~~~~~~~~~~~~~^^^
  File "/Users/nad/Projects/PyDev/active/dev/3x/source/Lib/json/decoder.py", line 345, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
               ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/nad/Projects/PyDev/active/dev/3x/source/Lib/json/decoder.py", line 363, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
make: *** [testios] Error 1
```

### Linked PRs
* gh-127592
* gh-127754
* gh-129124

### PR DESCRIPTION
**Pull request by freakboy3742**
Xcode 16 deprecated the `xcresultool` tool we were using to gather test results after an iOS test run. 

This tool was less than ideal anyway, as it didn't provide a way to see test results as the test suite was running. As a result, we had intermittent CI test failures that manifested as "test suite didn't finish", with no runtime debugging possible.

This PR adds a new configuration item (`config->use_system_logger`) for Apple platforms, providing the option to redirect stdout and stderr to the system log. This mirrors how Android operates - the implementation is strongly influenced by the Android implementation, but is just different enough that refactoring into a common base class is more trouble than it's worth IMHO.

This PR also adds a new test runner for iOS projects. This test runner uses the iOS testbed as a template, installing the compiled framework into copy generated from the template; and then building and running the test project in parallel with a log streamer that captures the system log from the iOS simulator as it runs. This allows us to observe test output as the test suite runs.

The test runner also includes options to allows the installation of app code. This means the testbed project can be used to test arbitrary Python code, in addition to the core CPython test suite.

Although this feature is primarily useful for iOS apps, the same code *can* be used for macOS apps - something that is useful when embedding Python in to a GUI app (where stdout/stderr isn't visible).

This PR resolves both #126925 *and* #126821; Resolving #126821 was a pre-requisite for fixing #126925 (and is the first commit in this PR), but by itself, there's very little opportunity to evaluate that it's working. You need an updated test runner to be able to observe that stdout/err redirection is working.

Fixes #126925
Fixes #126821

* Issue: gh-126925

----
📚 Documentation preview 📚: https://cpython-previews--127592.org.readthedocs.build/

### CODE REVIEW

**Comment by freakboy3742** [comment]
!buildbot iOS

**Comment by freakboy3742** [comment]
Moving to draft because of the test failures on macOS buils

**Comment by freakboy3742** [comment]
!buildbot iOS

**Comment by freakboy3742** [comment]
I've now resolved the testing issue; I've also added some documentation for the new capabilities.

File: Lib/test/test_apple.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
+import unittest
+from _apple_support import SystemLog
+from test.support import is_apple
+from unittest.mock import Mock, call
+
+if not is_apple:
+    raise unittest.SkipTest("Apple-specific")
+
+
+# Test redirection of stdout and stderr to the Android log.
```
Comment (ned-deily) [inline Lib/test/test_apple.py]: s/Android log/?/

**Review by ned-deily — APPROVED** [review]
A very impressive improvement: being able to follow the test run in real-time makes a world of difference. Congratulations! I did not make a thorough review especially of the details of how the logging would be configured for a Mac app. But I did verify that my standard build and execution of `make testios` runs correctly. I have two minor review comments. First, it would be better to condense the string of 30+ `Messages dropped` log entries to just one as the action to be taken is the same regardless of how many there are:

```
Filtering the log data using "senderImagePath ENDSWITH "/iOSTestbedTests.xctest/iOSTestbedTests" OR senderImagePath ENDSWITH "/Python.framework/Python""
Timestamp               Ty Process[PID:TID]
=== Messages dropped during live streaming (use `log show` to see what they were)
=== Messages dropped during live streaming (use `log show` to see what they were)
=== Messages dropped during live streaming (use `log show` to see what they were)
[...]

```

And a stray `Android` reference remains in a `test_apple.py` comment.

You might want to consider adding a `What's New` entry as well.

**Comment by freakboy3742** [comment]
> First, it would be better to condense the string of 30+ `Messages dropped` log entries to just one as the action to be taken is the same regardless of how many there are:

Agreed - this is a case where it's the iOS logger being messy, and this script is just returning the mess verbatim. However, it's something we might be able to clean up by checking for that specific content, and swallowing all but the first occurrence.

> And a stray `Android` reference remains in a `test_apple.py` comment.
...
> You might want to consider adding a `What's New` entry as well.

Ack - I'll add both these two.

**Comment by freakboy3742** [comment]
!buildbot iOS

**Comment by freakboy3742** [comment]
> I did not make a thorough review especially of the details of how the logging would be configured for a Mac app. 

There's no command line interface; it's currently only exposed as an option for embedded-mode apps. [Briefcase's stub binary](https://github.com/beeware/briefcase-macOS-Xcode-template/blob/main/%7B%7B%20cookiecutter.format%20%7D%7D/%7B%7B%20cookiecutter.class_name%20%7D%7D/main.m#L63), for example, would be able to add stdout redirection as part of the app startup sequence, rather than requiring the [`std-nslog`](https://pypi.org/project/std-nslog/) package to install a logging shim at runtime.

@ned-deily I've now made those changes, including surpassing all but the first occurrence of the "Messages dropped..." messages in any block when they occur. Did you want to review this again, or are you OK with me to merge and backport the changes?

**Comment by ned-deily** [comment]
LGTM, thanks

**Pull request merged**